### ISSUE TITLE
NFig Store Data

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#19]
Before settling on functional requirements for `NFigStore`, I want to figure out exactly what data the store needs to read/write. I'm going to use NFig.Redis as an example, since that's the first store that we'll need to implement.

### Key Format

All keys in NFig.Redis will follow a similar format: `NFig:{tier}:{keyName}:{appName}`

- __tier__: integer-representation of a deployment tier
- __keyName__: describes the type of data stored in the key (e.g. "Overrides" or "Metadata")
- __appName__: user-provided application name

The tier is included as a safeguard against reading/writing settings on the wrong tier. In some cases, the app name won't be applicable.

Example: `NFig:2:Overrides:MyApp`

At first glance, it might make more intuitive sense to reverse the `keyName` and `appName` positions. The problem is that the app name is user-provided, and could contain colons that we'd need to escape. It's easier to just put the app name at the end and not worry about escaping it.

## Global Data

Some data is not specific to an application.

### Application Names

A list of all applications which use the store for their settings. This will be stored as a Redis Set under the key `NFig:{tier}:AppNames`.

### Events

Events are indicators that some action has occurred. They are similar to logs, but with several key differences:

- Events are tracked globally, whereas logging works on an application level.
- Logs only track changes to overrides. Events are more broad. For example, updating metadata triggers an event, but is not logged.
- Events are ephemeral. Logs are stored indefinitely.

The data structure for an event should look something like:

```csharp
public class NFigEvent
{
    public Guid Guid { get; }
    public DateTimeOffset Timestamp { get; }
    public NFigEventType Type { get; }
    public string AppName { get; }
}
```

NFig.Redis will broadcast events on the `NFig:{tier}:Event` pub/sub channel. I haven't decided on a serialization format yet, but something human readable would be helpful.

NFig.Redis will also keep a rolling window of events in a Redis List named `NFig:{tier}:Events`. I'm thinking 100 events seems like a large enough rolling window for any real use case.

```
LPUSH NFig:{tier}:Events EVENT_DATA
LTRIM NFig:{tier}:Events 0 99
```

### Overrides

Stores all of the overrides, as well as the current commit ID for the app.

NFig.Redis will use a Redis Hash under the key `NFig:{tier}:Overrides:{appName}`. The commit will use the hash key `$commit`. Override values will use the hash key `{dataCenter}|{subAppId}:{settingName}` (subAppId will be omitted if it's not applicable).

This is the only Redis key that consuming apps will need to read.

### Sub-Apps

The names and IDs of every sub-app that an app client has declared. NFig.Redis will store these in a hash under `NFig:{tier}:SubApps:{appName}`. The keys of the hash will be the sub-app IDs, and the values will be the names of the sub-apps.

### Metadata

All metadata about settings which isn't subject to change during the lifetime of an app. This includes:

- Setting names, types descriptions, etc
- Default values
- Enum names and values for enums used in settings

This will be serialized as a JSON blob and stored in `NFig:{tier}:Metadata:{appName}`. The JSON should be consumable by the admin panel without transformation.

### Encryption Test

I'd like to provide some way for the admin panel to verify that it's using the same encryption method/keys as the app itself. The best way to do that seems like it would be to provide a sample encrypted value and test whether it can be decrypted. I'm thinking this could just be an encrypted version of the application name stored in `NFig:{tier}:EncryptedAppName:{appName}`.

### Logging

In principle, logging is just a history of changes to overrides. The serialization format of each log entry hasn't been entirely decided yet.

There are three different "views" on logs: by date, by user, and by setting. For a SQL store, this would be easy because you could simply add multiple indexes to the same table. For Redis, I think we're going to have to manually maintain three different copies using Redis Sorted Sets.

> We'll use _lexical_ ordering in the Sorted Sets, so the "score" will always be 0.

The keys will be:

- `NFig:{tier}:HistoryByDate:{appName}`
- `NFig:{tier}:HistoryByUser:{appName}`
- `NFig:{tier}:HistoryBySetting:{appName}`

Each will store the same values, but with a different prefix so that they're sorted differently.

- ByDate will use an ISO-8601 timestamp prefix. Example: `2017-05-14T20:16:26.7914969+00:00;LOG_RECORD`
- ByUser will use a user + timestamp prefix. The username will be normalized to lowercase. Example: `bcopeland:2017-05-14T20:16:26.7914969+00:00;LOG_RECORD`
- BySetting will use a setting name + timestamp prefix. Example: `Some.Setting.Name:2017-05-14T20:16:26.7914969+00:00;LOG_RECORD`

Sorted sets allow you to extract a portion of the set based on prefix. In most cases, descending order is what we would prefer. Redis supports descending order (`ZREVRANGEBYLEX`), but [StackExchange.Redis](https://github.com/StackExchange/StackExchange.Redis) currently does not. It's listed on the [2.0 milestone](https://github.com/StackExchange/StackExchange.Redis/milestone/1), so maybe we can prod at getting that to land.

**Comment by rossipedia** [comment]
Are we removing `TDataCenter` as a first class concept?

**Comment by bretcope** [comment]
No, data center is still part of the overrides. Everywhere else that it's applicable, such as in metadata or history entries, it's part of a value, rather than a key.

`TSubApp` on the other hand, is going away, but it's essentially just being replaced by an `int?`.

**Comment by rossipedia** [comment]
ahh, ok that makes sense

**Comment by bretcope** [comment]
I'm going to move default values out of metadata and into their own Redis hash under the key: `NFig:{tier}:DefaultsBySubApp:{appName}`. The hash keys will be the SubAppId, or `$root` in the case of the top-level application. The values will be a JSON dictionary, such as:

```
{
  "Some.Setting.Name": [
    { serialized_default_0 },
    { serialized_default_1 },
    ...
  ],
  ...
}
```

---

To explain the rationale, we need to talk about defaults for sub-apps. I definitely want to support setting a default for a specific sub-app:

```csharp
[Setting("some value")]
[SubAppDefault(MY_SUB_APP, "different value")]
public string Setting { get; }
```

Sub-apps are defined by ID. If that ID is a compile-time constant, then filling in `MY_SUB_APP` is no problem. However, that ID might _not_ be known at compile time.

For example, the SiteId of a Stack Exchange site lives in the Sites database, rather than a code enum. Since the ID is unknown at compile time, I would like to support setting a sub-app default by _name_.

```csharp
[Setting(false)]
[SiteDefault("Stack Overflow", true)]
public bool ShowJobBoard { get; }
```

Now, when NFig initially loads, it's not going to know how to map "Stack Overflow" to a sub-app, which means it really has no way to include that default in meta data.

When `AddSubApp()` is called to declare the Stack Overflow sub-app, then NFig knows how to generate the metadata for that sub-app. However, I don't want to constantly overwrite the metadata every time a sub-app is declared.

More importantly though, it is possible that some instances of an app might initialize different sub-apps. For example, some instances of the StackExchangeNetwork app might only initialize the "Meta Stack Exchange" sub-app, and it would be bad for it to overwrite the metadata with _only_ defaults for Meta.

The easiest way to fix all of these issues is to store the defaults for each sub-app as a distinct Redis entry.

**Comment by captncraig** [comment]
The encryption test is very interesting to me. It kinda depends on the security model. Is the asymmetric mode intended to let the admin panel write vales with a public key, but not read them? If so, the underlying application could supply the public key via the data store (possibly per app or even per tier), and the admin interface uses that to encrypt values. 

With symmetric encryption, it gets harder to verify the keys match. Are you just supplying the same key to both the app and the data store independently?Are you suggesting the app encrypt a well know value, just so the admin ui knows it can read it properly? 

What is the threat model? Just hiding values from people with ui access? or maybe from someone with direct access to redis? Or something else?

**Comment by bretcope** [comment]
Interesting questions. How to deliver the private key is up to the owner of the app. I don't think NFig should put it in the persistent store, though an app could certainly do that manually if they really wanted. It's less about who has access to Redis (or whatever backing store), and more that I think people would be very surprised if we saved their private key somewhere without them explicitly doing so. Surprises are generally a bad thing when it comes to security.

I don't think we've settled on a private key distribution method at Stack. I know we've discussed TeamCity, but there are other options.

In some circumstances, a _remote_ admin panel may not be able to decrypt _or_ encrypt settings. It would be the responsibility of the user to pre-encrypt any overrides they want to set. It certainly makes the admin panel less useful, but I don't want to dictate too much in terms of security.

Now, the public key is an interesting aspect I hadn't thought about before. There's no reason it couldn't be persisted if it were readily available from the `ISettingsEncryptor` (not currently the case). I could maybe see a use case for a remote admin panel where you can set, but not read, encrypted settings. I don't think it's something we would do, but someone might.
